### ISSUE TITLE
Needs a way to set a user's voice mode. (In flux or components)

### ISSUE THREAD

**Issue by AmasterAmaster** [Yellow-Dog-Man/Resonite-Issues#2017]
### Is your feature request related to a problem? Please describe.

Currently there is only a way to get a user's voice mode, but no easy way of setting it directly.

My use case for this is that I am currently creating a classroom world, and the instructor (whoever it may be) should be able to press a button to change their current voice mode to Broadcast (and when pressed again should be able to return to the previous stored voice mode). This at first sounds trivial, as the instructor would normally just use the dash facet to change it, but the following example is truly why I would like this feature...

For the students sitting in the seats in this classroom, there is a button provided in front of them that will act the same way, pressing it should set their voice mode to broadcast so their questions and concerns can be heard for the instructor/students in the room. Now, the instructor can't really enter the other user's dashes to turn on broadcast mode for them, so a flux node would be best here to let the button change the local user's voice mode for them (making it easy for everyone in this scenario).

### Describe the solution you'd like

I would like to see a flux node or component that can allow for setting a user's voice mode.

### Describe alternatives you've considered

I have considered the following:
 - Having the instructor ask the other users to kindly change their voice mode to broadcast (but that is inconvenient and slow, and new users may not know how to do this), and also having to ask them to change it back, having this happen every time a person has a question can make the whole experience slow down to a crawl, which is definitely not ideal.
 - Having the instructor open a User Inspector and changing it manually, however this is more invasive and not really a clean way of changing a voice mode.
 - Using ref hacking to access the user's voice mode from a button press, this would be a hacky way of doing it and it would be better if there was a node that could set it directly instead of trying to find the reference that holds it.

### Additional Context

This can be used for other contexts as well, for some examples:
 - What if a world creator wanted to make a game show and have a button press to make the user's voice turn into broadcast mode for everyone to hear the answer to something.
  - What if a world creator wanted to make a presentation world and the people on stage holding microphones need to be in broadcast mode for everyone to hear them (letting the grabbed prop set the voice mode, and reverting it back when it is no longer grabbed).

### Requesters

AmasterAmaster

**Comment by epicEaston197** [comment]
I know this issue is made with the best of intentions but this could easily be abused

**Comment by AmasterAmaster** [comment]
@epicEaston197 Like with anything that is user created content, anything can be abused. and following from the advise I heard, if someone is abusing something, file a ticket on it. I also follow the mantra of "A tool is a tool, it can be used for good or bad, but at the end of the day, it is just a tool".

**Comment by epicEaston197** [comment]
> @epicEaston197 Like with anything that is user created content, anything can be abused. and following from the advise I heard, if someone is abusing something, file a ticket on it. I also follow the mantra of "A tool is a tool, it can be used for good or bad, but at the end of the day, it is just a tool".

Yeah but I still don't particularly feel comfortable with someone being able to unmute me without my consent

**Comment by ohzee00** [comment]
Does GrabbableUserVoiceModifier not follow your usecases Amaster? I haven't tested it in a long time but it should work on grabbable objects to change a user's voice to say, broadcast when grabbing a certain object.

**Comment by epicEaston197** [comment]
> Does GrabbableUserVoiceModifier not follow your usecases Amaster? I haven't tested it in a long time but it should work on grabbable objects to change a user's voice to say, broadcast when grabbing a certain object.

That component doesn't appear to function

**Comment by AmasterAmaster** [comment]
@epicEaston197 True, that could be a thing the devs could do, like a stronger mute option that prevents force unmuting, also the world creator should take into account of that, but I understand that people may not know nor have the foresight to this, thus the setting would override this node/component. Also, if privacy is that much of a concern, I would think that users would rather just open a clean grid world (with no background flux of any kind) and talk privately there as well, there are options.

**Comment by epicEaston197** [comment]
> @epicEaston197 True, that could be a thing the devs could do, like a stronger mute option that prevents force unmuting, also the world creator should take into account of that, but I understand that people may not know nor have the foresight to this, thus the setting would override this node/component. Also, if privacy is that much of a concern, I would think that users would rather just open a clean grid world (with no background flux of any kind) and talk privately there as well, there are options.

Muting my microphone should not be treated like the whisper bubbles muting my microphone should be a privacy feature and it should be respected

**Comment by AmasterAmaster** [comment]
@epicEaston197 Yes, hence why I mentioned that being a setting they could setup, which would block this from being a feature, but once that is setup, then world creators can finally have a way to structure their classes in different ways (say for example, if people are to take a test, then make all users mute until the exam is done). Then the instructor should have that control, plus it is implied that if your in a classroom, then you would at least be communicating to the teacher/students in some capacity (muted or not). But that is its own moral dilemma (And I am thinking in more of a grander scale of separate Resonite universes that may need that type of control for their students).

**Comment by Zyzyl** [comment]
This would be useful. I can imagine situations in game worlds or Metamovies where this functionality would be great (e.g. simulating switching to radio comms by setting people to Broadcast).

It's worth noting that there it's already possible to 'Mute' or 'Silence' users at a session/world level which doesn't override their own voice level set in the Dash. I generally agree that there should be protections against being able to fully unmute another user without their consent. This is how it works in Zoom for example, the call Host can mute people, but can only invite them to unmute themselves.

**Comment by JackTheFoxOtter** [comment]
> Does GrabbableUserVoiceModifier not follow your usecases Amaster? I haven't tested it in a long time but it should work on grabbable objects to change a user's voice to say, broadcast when grabbing a certain object.

That thing has been broken for ages, but yeah it would be awesome to see that fixed.

**Comment by ohzee00** [comment]
> 
> That thing has been broken for ages, but yeah it would be awesome to see that fixed.

Not trying to steal this issue but doin' a separate GitHub issue on that would be good, I didn't realize it was broken honestly. (And I'm gonna sleep so I'm totally going to forget to make one)

**Comment by shadowpanther** [comment]
Couldn't cloud vars and a Dash facet (on each consenting user) be used to do this?

I.e., make a facet that has a voice mode switcher in it, and has flux on it that changes voice mode when the cloud var is changed. Then the objects in the world would be set to write users' cloudvars.

**Comment by Chundelac** [comment]
> > @epicEaston197 Like with anything that is user created content, anything can be abused. and following from the advise I heard, if someone is abusing something, file a ticket on it. I also follow the mantra of "A tool is a tool, it can be used for good or bad, but at the end of the day, it is just a tool".
> 
> Yeah but I still don't particularly feel comfortable with someone being able to unmute me without my consent maybe an additional privacy setting could be added to prevent this from working

I believe it is already possible to force unmute someone.

**Comment by Dusty-Sprinkles** [comment]
> Couldn't cloud vars and a Dash facet (on each consenting user) be used to do this?
> 
> I.e., make a facet that has a voice mode switcher in it, and has flux on it that changes voice mode when the cloud var is changed. Then the objects in the world would be set to write users' cloudvars.

If this is targeting newer users who have a hard time figuring out broadcast imagine how hard teaching each of them individually to install a facet would be.

Also yea I'm for this or fixing the component, or both, it would be nice to have for a lot of applications and maybe just some settings if you don't want to be force unmuted/unwhispered/muted should be enough imo

**Comment by Psychpsyo** [comment]
I agree that this would be a nice thing but there definitely should not be a way to force-unmute somebody else.

Oftentimes the fact that someone is muted means that they're having some form of private conversation with someone IRL or on a call.
As for the "anything can be abused because it's user-created content":
Certain forms of abuse are unavoidable, temporary, 'not that bad' or easily moderated.
Like crashing someone, lagging their game, loud noises, flashing lights, etc.

Random mic access is not one of those.
It is avoidable, not a temporary problem if someone makes you leak private data, quite bad depending on what's leaked and you can't moderate the problem away after it's been abused and someone get the info they wanted.

**Comment by AmasterAmaster** [comment]
@Psychpsyo That is why there should be the exception to the node to either not force unmute (limiting it to a few voice modes) OR have a setting that prevents force unmuting for users, especially if a world creator wanted to have access to all voice modes. (I am in favor of the setting option).

As for your second point, I still say having a node like this would be considered a tool anyone can use, and yea someone would probably use it maliciously, but I feel it has a greater good, like accessibility for example (new users that are just starting would not know how to use voice mode controls or maybe have not noticed them, but also would want to learn something in a classroom setting and would happily take advantage of a button/toggle near them to help them speak up and to be heard from a long distance).

I really like your suggestion of a requesting feature though, it could act similarly to what we already have for opening hyperlinks and webpages, just having a popup saying the reason why would be perfect!

**Comment by epicEaston197** [comment]
> Can you please report this as its own issue, @epicEaston197 @ohzee00?

I've made [#2087](https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/2087)

**Comment by PraxThing** [comment]
> You can also already change a user's voice mode through the user inspector.

Voice mode should be read-only unless there is explicit consent to have it changed, by any means other than the user's own dash/facets.  This consent should be per-session and ephemeral.

Access to changing anything about the voice mode's _behavior_ should be restricted;   Either changing the voice mode or altering the distance at which you can be heard are privacy-affecting operations and should require consent.  For example, there is an expectation that the whisper bubble accurately represents the the range at which someone can be heard.  Another user with an inspector / session permissions should not be able to circumvent that without some kind of prompt to the affected user.

**Comment by Nytra** [comment]
If a user is muted via the dash, it is not possible to unmute them through the user inspector.

**Comment by Frooxius** [comment]
@PraxThing That depends a lot on the session. There are worlds/cases where the host controls the mode of users - e.g. game modes, where any "active" user is automatically put to shout/broadcast.

If you're already streaming your voice into the session, it's considered "free game".

The only exception to this is when you're muted - we mute at the driver level outside of the world itself, meaning the actual audio stream stops - there's nothing the world can do to unmute you in that case.

It's also important to note that whisper bubbles are NOT a privacy feature, they're an accessibility one to prevent people getting overwhelmed in crowded/busy/loud worlds.

Even if the voice mode is not changed, there's lots of ways for someone to listen in - they become tiny and enter the voice bubble, they hook into the audio stream (this can be culled with future updates, but that doesn't resolve issue fully) or even the host can hook into that, since they're relaying the data. Fundamentally, they're not a privacy feature.

For private conversations, we recommend starting a new session which has only people you trust.

**Comment by AmasterAmaster** [comment]
In this scenario, I have bad hearing, and when I give a lecture and teach things in a classroom world, I am unlikely to stop if I don't hear anyone speaking to me or from behind me (when I am facing away especially when using earmuff mode). Having a simple node that controls the voice mode (hooked up to a button) can allow for any person to change to broadcast mode and get my attention easily and quickly.

Safety and implementation is on the dev's side, and I am in favor of safety features for this node (accounting for muting and things like that). But having a feature like this will also help prevent silly ways to getting the voice mode such as ref hacking or using another avatar to grab the audio reference (making things more efficient). I would appreciate this addition at least.

**Comment by Nytra** [comment]
At the moment you can inject a AvatarAudioConfiguration component to each user to make the default voice range higher, and provide a button with a virtual M key to toggle the user between a muted and unmuted state

**Comment by badhaloninja** [comment]
> At the moment you can inject a AvatarAudioConfiguration component to each user to make the default voice range higher, and provide a button with a virtual M key to toggle the user between a muted and unmuted state. You could even drive the audio configuration with dynamic variables which you could hook up to buttons if you like.

I made an example of this for the MMC23 Kickoff Theatre (`resrec:///G-Creator-Jam/R-51ef72e4-7dd5-4251-a981-86d528ac4de4`), iirc host should have Admin perms
I don't remember if we reused it for the MMC24 Theatre

This was used for making people on stage Shout automatically so they aren't too quiet when on camera
You can find it under `The World > Logix > Shout on stage bit > AutoInject` (sorry about the world hierarchy)
This system works by toggling different data presets depending on collider interactions
